### ISSUE TITLE
Do not call sys.exc_clear() on Python 3 in bin/admin

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8465]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8465|
|Type|enhancement|
|Created|2016-06-13 21:59:28Z|
|Branch|https://github.com/twisted/twisted/tree/8465-rodrigc-exc_clear-2|

sys.exc_clear() is gone on Python 3: https://docs.python.org/3/whatsnew/3.0.html#index-22

We should not try to call it on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__8465 8465
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8465_rodrigc_exc_clear_2 8465-rodrigc-exc_clear-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1465855168919604 1465855168919604
changetime__1483365140930117 1483365140930117
version__None None
owner__rodrigc rodrigc

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_1"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/184

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

Ok, at a high level this looks ok so I'm sending it to the builders. More review coming.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_3"></a>|@Lukasa set owner to @rodrigc|
|-|-|

Ok, I'm letting the builders run. In the mean time, here's some review that applies not only to this but to [#8466](https://github.com/twisted/twisted/issues/8466), [#8467](https://github.com/twisted/twisted/issues/8467), [#8468](https://github.com/twisted/twisted/issues/8468), and [#8469](https://github.com/twisted/twisted/issues/8469): what the hell is `exc_clear()` even doing there? It seems to serve no purpose whatsoever. Glyph has suggested it has something to do with a Python 2.2 implementation detail, so I'm inclined to say this is crazy-talk.

I'm going to let this build run, but I'd like to see one of two things:

- Either more aggressive patch that simply changes the except block to `pass`.
- Or the most aggressive patch, which removes all of this nonsense and makes these all `console_scripts`, such that `pip install -e '.[dev]'` (the command we tell people to run for development anyway) makes those scripts available directly.

I'll happily accept either kind of patch, but we should definitely do one of them.

I'll apply this feedback to the other tickets as well. Feel free to ping me directly when you're ready for more review.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_4"></a>|@rodrigc set owner to @Lukasa|
|-|-|

I updated the branch here: https://github.com/twisted/twisted/compare/trunk...8465-rodrigc-exc_clear

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/184

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_7"></a>|@Lukasa commented|
|-|-|

Cool, eyeball review looks good. TO THE BUILDERS!

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_8"></a>|@Lukasa commented|
|-|-|

twistedchecker is misbehaving on this patch too, but none of the failures apply here so we're going to disregard it.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_9"></a>|@Lukasa set owner to @rodrigc|
|-|-|

A beautiful wall of green! You're good to merge this.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|GitHub set status to `closed`|
|-|-|

In changeset 7459bde888a93f7787da92ff34b8c207c695a2c9
```
#!CommitTicketReference repository="" revision="7459bde888a93f7787da92ff34b8c207c695a2c9"
Merge 8465-rodrigc-exc_clear: Do not call sys.exc_clear() on Python 3 in bin/admin

Author: rodrigc
Reviewer: Lukasa
Fixes: #8465
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

The exc_clear call was important for starting application code (in the case of trial and twistd, primarily) without an exception state (the _preamble ImportError) set in the execution context.

This is moot now - but a little bit of searching through the repository history should have revealed the motivation.
